### ISSUE TITLE
Enhancement: Support Hashbang parse in `typescript-estree`

### ISSUE THREAD

**Issue by fisker** [typescript-eslint/typescript-eslint#6500]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

I understand that ESLint core handles Hashbang automatically, but it's kind of missing feature when use `typescript-estree` alone.

Should we add support for it? It's in the spec after all.

### Fail

```typescript
N/A
```

### Pass

```typescript
N/A
```

### Additional Info

```js
> require('@typescript-eslint/typescript-estree').parse('#!/usr/bin/env node', {comment: true})
{ type: 'Program', body: [], sourceType: 'script', comments: [] } // <- missing comment
```

**Comment by bradzacher** [comment]
Do any other parsers handle it?
It looks like other estree-compatible parsers just ignore it.

**Comment by fisker** [comment]
[acorn](https://github.com/acornjs/acorn/commit/e708a22874739825a3ed87eb3b265eadb8d6b947) set `allowHashbang` to `ture` when using latest ecma version.

[espree](https://github.com/eslint/espree/pull/556).

meriyah also support it.

**Comment by bradzacher** [comment]
The reason we haven't historically supported it is because ESLint will do some slightly dodgy stuff in a way and replace the `#!` with `//` so that parsers didn't need to support it.

Then after the parse ESLint checks to see if the original file starts with `#!`, then updates the first comment to `type = 'Shebang'`.

espree parses it out as `type = 'Hashbang'`, which what I guess we'll do too considering they chose hashbang as the name in the spec.

**Comment by fisker** [comment]
I'll work on this.

**Comment by JoshuaKGoldberg** [comment]
Per discussion in https://github.com/typescript-eslint/typescript-eslint/pull/6732#issuecomment-1630720900, we'd want ESLint core to support it for us to. Marking this as blocked on ESLint.

Interestingly, https://github.com/eslint/eslint/issues/16158#issuecomment-1194700785 confirms what Brad noted about ESLint replacing with `//`.

cc @bradzacher - I think what I'm doing is what your comments lead to?

**Comment by Josh-Cena** [comment]
> so hopefully a newer ECMA Version will support it?

ECMA or ESLint? This feature is already in ES2023.

**Comment by bradzacher** [comment]
I'm not sure if ESLint core will ever do this given they have "shebang" comments.
Doing this would break some old rules due to the `type` change.
Maybe we could push on it to be a v9 breaking change for them?

**Comment by JoshuaKGoldberg** [comment]
> > so hopefully a newer ECMA Version will support it?
> 
> ECMA or ESLint? This feature is already in ES2023.

Heh, right, ESLint. Now that ES2023 has it I would think ESLint's parser would need to respect the changes.

@fisker are you up for filing an issue on ESLint?

**Comment by fisker** [comment]
Sorry, I don't really understand why we can't have different types for the parser and the ESLint plugin.

I don't want push ESLint to change their AST type either, since I have rules maintained, it will break other ESLint community plugins too. I don't want be the "bad" guy.

Anyway, change it to match the spec make sense too.  I can live with that if they decide to make the change.

**Comment by JoshuaKGoldberg** [comment]
Makes sense. I'm doing a bit of a deep dive now. The discussion in https://github.com/estree/estree/pull/224 -> https://github.com/estree/estree/pull/224#issuecomment-640771622 is interesting and relevant:

> Closing this as hashbang is considered a comment and thus should not be included in estree.

https://github.com/estree/estree/issues/286 was closed as a duplicate after hashbang grammar hit stage 4.

Anyway, if ESLint is treating it as a comment then I think we could at least put it in the comments as a _Line_ comment. So, not quite what #6732 has currently implemented with a new _Hashbang_ comment, but still adding it to the `comments: []` mentioned in the OP.

cc @bradzacher

**Comment by JoshuaKGoldberg** [comment]
So: an explicit opt-in option to allow parsing a hashbang as a line comment, then?

**Comment by fisker** [comment]
I'm fine to add option, Prettier only need write parseOptions once, 😄 


If we add a new "Shebang" to the type, when using in eslint plugin, the AST is actually `"HashBang"`, will that be a problem? I'm not familiar with TS, but is it possible to define different AST type for the two parsers(estree and parseForEslint)? 

https://github.com/typescript-eslint/typescript-eslint/pull/6732/files#diff-bb2be6d717f000780e889eb989aeceab7bc408c1a1c4041b721781cdf95ffa6fL39

**Comment by bradzacher** [comment]
My big issue is the types - we can't correctly type this for our consumers with an option.

I know this isn't a concern for prettier - but it is a concern for other consumers.

We can't just add the new comment to the `Comments` union type because that'll break the linting usecase (which well see "shebang" type comments, not "hashbang" comments) and I don't know if we can conditionally modify the types either.

If we can solve the types without breaking the linting usecase then I'm happy for it to be added.

**Comment by Josh-Cena** [comment]
Sorry, I'm not fully up to speed with the discussion but I find it interesting. What is the break we are trying to avoid? I'm also confused by what ESLint does: what comment type does a plugin see and what type does the core see, and why does it matter? Does ESLint call the comment type "Shebang" instead of "Hashbang"?

**Comment by bradzacher** [comment]
ESLint supports hasbangs via some (hacky) logic that is essentially:
1) load file from disk
2) if file starts with `#!`, replace with `//`
3) pass to parser
4) if (2) matched, update the first comment's type -- `ast.comments[0].type = 'Shebang'`
5) lint the AST

They did this because way back when hashbang wasn't standardised so most parsers didn't support it.

So that means from the linting side of things -- if we did add handling for this then lint rules would never see the AST node because ESLint would never give us a hashbang comment.
So we'd declare our types as `type Comment = LineComment | BlockComment | HashBangComment` but the latter would be an unreachable union member.

The only time it'd be reachable is via custom usages of `typescript-estree` (i.e. the prettier usecase).

So we're in a pickle - if we add support for it then we have a branch that's unreachable from the most common usecase. Branching the types is complicated, but doable.

I personally like the idea of supporting it - but I just don't like the type problem it produces.
We could handle it -- or we could wait until ESLint itself fixes things to add support properly.

**Comment by Josh-Cena** [comment]
I personally don't see a downside other than having a "generally" (though not always) unreachable branch. Any plugin continues to work, the ESLint core continues to work, and this will only make external parser consumers' lives easier. Furthermore this means we probably don't need to do any extra work when ESLint one day passes the `#!` to us verbatim.

**Comment by bradzacher** [comment]
> other than having a "generally" (though not always) unreachable branch

In lint rules it would always be unreachable.

As per the new spec the comment is called a "Hashbang" - they purposely did not call it "Shebang" - and so the node must be have the type `Hashbang`
However ESLint types the node as a `Shebang`.

So in lint rules you have:
```ts
switch (comment.type) {
  case 'Line': // ...
  case 'Block': // ...
  case 'Shebang': // ...
  case 'Hashbang': throw new Error("Always Unreachable");
}
```

Whereas from the direct `typescript-estree` consumer you have the opposite with
```ts
  case 'Shebang': throw new Error("Always Unreachable");
  case 'Hashbang': // ...
```

**Comment by Josh-Cena** [comment]
Hmm. I see the point. Well, won't ESLint switching from `Shebang` to `Hashbang` always be a breaking change, then? Doesn't that mean `Shebang` will stick forever in estree and we should just emit that instead of `Hashbang`?

**Comment by bradzacher** [comment]
It would be a breaking change for them - so unsure of when or if they'll ever do it.

If we want to truly be spec compliant then we technically need to support `Hashbang`.
I think that ESLint has been punting on it because it's easier to ignore it..
It's probably worth asking them if there are plans. I'd guess the answer is "maybe after v9" though.

**Comment by bradzacher** [comment]
> With which spec?

The ESTree spec - which doesn't technically spec comments IIRC - but the parsers that follow estree all use the same comment spec. Cos if they don't then a lot of lint rules won't work!

**Comment by Josh-Cena** [comment]
So, if I understand correctly: ESTree-compliant parsers emit `Hashbang` but ESLint emits `Shebang`? Is that the current divergence?

**Comment by fisker** [comment]
If we change `@typescript-eslint/parser`.`parseparseForESLint` and `@typescript-eslint/typescript-estree`.`{parse,parse*}` to differnt result, this can be fixed, right?
